packages/core/src/memo.ts

```diff
@@ -0,0 +1,29 @@
+import type { StateNode } from '.';
+
+const stateNodeMap = new WeakMap<StateNode, any>();
+
+export function getMemo<T>(
+  stateNode: StateNode<any, any>,
+  key: string
+): T | undefined {
+  return stateNodeMap.get(stateNode)?.[key];
+}
+
+export function memo<T>(
+  stateNode: StateNode<any, any>,
```

Comment:
this is quite a generic module - we could decouple its implementation from state nodes, the "owner" of the memo cache can be just anything (well, anything that is accepted by a `WeakMap`), it's not limited to state nodes